File: 1136-parallel-courses/1136-parallel-courses.cpp
class Solution {
public:
    int minimumSemesters(int n, vector<vector<int>>& relations) {
        vector<vector<int>> adj(n+1); 
        vector<int> indegree(n+1); 
        for(vector<int>& v : relations){
            int from = v[0], to = v[1]; 
            adj[from].push_back(to); 
            indegree[to]++; 
        }

        queue<int> q; 
        for(int i = 1; i <= n; i++){
            if(indegree[i] == 0){
                q.push(i); 
            }
        }

        int time = 0; 
        int nodes = 0; 
        while(!q.empty()){
            int size = q.size();
            time++; 
            for(int i = 0; i < size; i++){
                int node = q.front();
                q.pop(); 
                nodes++; 
                for(int next : adj[node]){
                    indegree[next]--; 
                    if(indegree[next] == 0){
                        q.push(next); 
                    }
                }
            } 
        }
        //cout << nodes << ' ' << time; 
        return nodes == n ? time : -1; 
    }
};

File: 1136-parallel-courses/1136-parallel-courses.java
class Solution {
    public int minimumSemesters(int n, int[][] relations) {
        Queue<Integer> q = new LinkedList<>(); 
        int[] inDegree = new int[n+1]; 
        List<List<Integer>> adj = new ArrayList<>(); 
        for(int i = 0; i <= n; i++){
            adj.add(new ArrayList<>()); 
        }
        for(int[] p : relations){
            int prev = p[0]; 
            int next = p[1]; 
            adj.get(prev).add(next); 
            inDegree[next]++; 
        }

        for(int i = 1; i <= n; i++){
            if(inDegree[i] == 0){
                q.offer(i); 
            }
        }

        if(q.isEmpty()) return -1; 
        int term = 0; 
        while(!q.isEmpty()){
            int size = q.size();
            term++;  
            for(int i = 0; i < size; i++){
                int curr = q.poll(); 
                for(int nextNode : adj.get(curr)){
                    inDegree[nextNode]--;
                    if(inDegree[nextNode] <= 0){
                        q.offer(nextNode); 
                    }
                }
            }
        }

        return term; 
    }
}